From: Michael Ellerman <michael@ellerman.id.au>
Subject: [PATCH] powerpc: iseries: Add /system-id, /model and /compatible
X-Git-Commit: 584fc6d111c34a9a2512f6c7652dff29232bf70d
X-Git-Commit: 289f1c74a9b527a816f63092e79a5412cc0352a2
Patch-mainline: 2.6.17-rc1

Add /system-id, /model and /compatible to the iSeries device tree.
Add strne2a() which converts a string from EBCDIC to ASCII.

Signed-off-by: Michael Ellerman <michael@ellerman.id.au>
Signed-off-by: Paul Mackerras <paulus@samba.org>
Signed-off-by: Olaf Hering <olh@suse.de>

 arch/powerpc/lib/e2a.c                 |   10 ++++++++++
 arch/powerpc/platforms/iseries/setup.c |   20 ++++++++++++++++++++
 include/asm-powerpc/system.h           |    2 ++
 3 files changed, 32 insertions(+)

Index: linux-2.6.16/arch/powerpc/lib/e2a.c
===================================================================
--- linux-2.6.16.orig/arch/powerpc/lib/e2a.c
+++ linux-2.6.16/arch/powerpc/lib/e2a.c
@@ -105,4 +105,14 @@ unsigned char e2a(unsigned char x)
 }
 EXPORT_SYMBOL(e2a);
 
+unsigned char* strne2a(unsigned char *dest, const unsigned char *src, size_t n)
+{
+	int i;
 
+	n = strnlen(src, n);
+
+	for (i = 0; i < n; i++)
+		dest[i] = e2a(src[i]);
+
+	return dest;
+}
Index: linux-2.6.16/include/asm-powerpc/system.h
===================================================================
--- linux-2.6.16.orig/include/asm-powerpc/system.h
+++ linux-2.6.16/include/asm-powerpc/system.h
@@ -171,6 +171,8 @@ extern u32 booke_wdt_period;
 
 /* EBCDIC -> ASCII conversion for [0-9A-Z] on iSeries */
 extern unsigned char e2a(unsigned char);
+extern unsigned char* strne2a(unsigned char *dest,
+		const unsigned char *src, size_t n);
 
 struct device_node;
 extern void note_scsi_host(struct device_node *, void *);
Index: linux-2.6.16/arch/powerpc/platforms/iseries/setup.c
===================================================================
--- linux-2.6.16.orig/arch/powerpc/platforms/iseries/setup.c
+++ linux-2.6.16/arch/powerpc/platforms/iseries/setup.c
@@ -50,6 +50,7 @@
 #include <asm/iseries/hv_call_xm.h>
 #include <asm/iseries/it_lp_queue.h>
 #include <asm/iseries/mf.h>
+#include <asm/iseries/it_exp_vpd_panel.h>
 #include <asm/iseries/hv_lp_event.h>
 #include <asm/iseries/lpar_map.h>
 #include <asm/udbg.h>
@@ -917,6 +918,24 @@ void dt_cpus(struct iseries_flat_dt *dt)
 	dt_end_node(dt);
 }
 
+void dt_model(struct iseries_flat_dt *dt)
+{
+	char buf[16] = "IBM,";
+
+	/* "IBM," + mfgId[2:3] + systemSerial[1:5] */
+	strne2a(buf + 4, xItExtVpdPanel.mfgID + 2, 2);
+	strne2a(buf + 6, xItExtVpdPanel.systemSerial + 1, 5);
+	buf[11] = '\0';
+	dt_prop_str(dt, "system-id", buf);
+
+	/* "IBM," + machineType[0:4] */
+	strne2a(buf + 4, xItExtVpdPanel.machineType, 4);
+	buf[8] = '\0';
+	dt_prop_str(dt, "model", buf);
+
+	dt_prop_str(dt, "compatible", "IBM,iSeries");
+}
+
 void build_flat_dt(struct iseries_flat_dt *dt, unsigned long phys_mem_size)
 {
 	u64 tmp[2];
@@ -927,6 +946,7 @@ void build_flat_dt(struct iseries_flat_d
 
 	dt_prop_u32(dt, "#address-cells", 2);
 	dt_prop_u32(dt, "#size-cells", 2);
+	dt_model(dt);
 
 	/* /memory */
 	dt_start_node(dt, "memory@0");
